Topic/cuda aware communications

bosilca opened the pull request:
Add support for sending and receiving the data directly from and to devices. There are few caveats (noted on the commit log).

Note: because it includes the `span` renaming, this PR changes teh public API and will need to bump version to 5.x


1. The first question is how is such a device selected ?

The allocation of such a copy happen way before the scheduler is invoked
for a task, in fact before the task is even ready. Thus, we need to
decide on the location of this copy only based on some static
information, such as the task affinity. Therefore, this approach only
works for owner-compute type of tasks, where the task will be executed
on the device that owns the data used for the task affinity.

2. Pass the correct data copy across the entire system, instead of
   falling back to data copy of the device 0 (CPU memory)

TODOs

* [x] rebase on c11 atomic fix
* [x] Add a configure option to enable GPU-aware communications.
* [x] Add a runtime configuration to turn on/off the gpu-aware comms? 
* [x] Pass -g 2 tests
* [x] Failure with ctest get_best_device `scheduling.c:157: int __parsec_execute(parsec_execution_stream_t *, parsec_task_t *): Assertion NULL != copy->original && NULL !=      copy->original->device_copies[0]'`
* [x] Failure with ctest nvlink, stress (segfault), details of why (its because using NEW) https://github.com/ICLDisco/parsec/pull/671#issuecomment-2588223287
* [x] Failure with ctest stage (presumably identical to intermittent failure in gemm/potrf) `device_gpu.c:2470: int parsec_device_kernel_epilog(parsec_device_gpu_module_t *, parsec_gpu_task_t *): Assertion PARSEC_DATA_STATUS_UNDER     _TRANSFER == cpu_copy->data_transfer_status' failed.`
* [x] RO data between tasks may reach an assert when doing D2D between devices that do not have peer_access between them (relevant only for weirdos)
* [x] POTRF Crashes on Frontier (due to using uninitialized/after free device_copies  https://github.com/ICLDisco/parsec/pull/671/commits/8976f0b00bdc375171b4d4bd07f39e5744773370
* [x] readers values are miscounted when 2 or more GPUs are used per rank https://github.com/ICLDisco/parsec/pull/671#issuecomment-2408016948)

bosilca committed:
Add a CUDA-based RTT test.

This allows to check if the data can be send and received directly to
and from GPU buffers.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Transfer data to and from GPU.

This is a multi-part patch that allows the CPU to prepare a data copy
mapped onto a device.

1. The first question is how is such a device selected ?

The allocation of such a copy happen way before the scheduler is invoked
for a task, in fact before the task is even ready. Thus, we need to
decide on the location of this copy only based on some static
information, such as the task affinity. Therefore, this approach only
works for owner-compute type of tasks, where the task will be executed
on the device that owns the data used for the task affinity.

2. Pass the correct data copy across the entire system, instead of
   falling back to data copy of the device 0 (CPU memory)

Add a configure option to enable GPU-aware communications.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Allow JDF with no dependencies, no datatype and no arenas.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Mostly improvement to the debuging output.

Name the data_t allocated for temporaries allowing developers to track
them through the execution. Add the keys to all outputs (tasks and
copies).

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

abouteiller commented:
This crashes in dpotrf, an original used in debug outputs during RELEASE_DEP_OUTPUT (generated code for T) is not initialized (neither NULL nor valid). 

```
"testing_dpotrf" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe757fa000 (LWP 1529612)]
parsec_task_snprintf (str=0x7ffe757ef0a0 "potrf_dtrsm(0, 1)[0, 1[, 1][, 5]]<45> keys = {*", size=128, task=0x7ffe757ef200) at /home/bouteill/parsec/dplasma-master/parsec/parsec/parsec.c:1952
1952                index += snprintf(str + index, size - index, "%s%lx", prefix, task->data[i].data_in->original->key);
(gdb) bt
#0  parsec_task_snprintf (str=0x7ffe757ef0a0 "potrf_dtrsm(0, 1)[0, 1[, 1][, 5]]<45> keys = {*", size=128, task=0x7ffe757ef200) at /home/bouteill/parsec/dplasma-master/parsec/parsec/parsec.c:1952
#1  0x00007ffff56e6b03 in iterate_successors_of_dpotrf_U_potrf_dpotrf (es=0x7ffe5c000bf0, this_task=0x107d990, action_mask=989855747, ontask=0x7ffff4218207 <parsec_set_up_reshape_promise>, ontask_arg=0x7ffe757ef6a0)
    at /home/bouteill/parsec/dplasma-master/build.hip/src/dpotrf_U.c:6919
#2  0x00007ffff56e76a5 in release_deps_of_dpotrf_U_potrf_dpotrf (es=0x7ffe5c000bf0, this_task=0x107d990, action_mask=989855747, deps=0x0) at /home/bouteill/parsec/dplasma-master/build.hip/src/dpotrf_U.c:7046
#3  0x00007ffff56e8f84 in complete_hook_of_dpotrf_U_potrf_dpotrf (es=0x7ffe5c000bf0, this_task=0x107d990) at /home/bouteill/parsec/dplasma-master/build.hip/src/dpotrf_U.c:7801

(gdb) p * task->data[i].data_in->original                                                                                                                                                                                                 
Cannot access memory at address 0xeda034a1
```

abouteiller commented:
On Frontier, there is a problem with the initializer for `  gdata->lock               = PARSEC_ATOMIC_UNLOCKED;` in `zone_malloc_init`, it doesn't compile. 

```
[  0%] Building C object parsec/parsec/CMakeFiles/parsec-base-obj.dir/utils/zone_malloc.c.o
/ccs/home/bouteilla/parsec/dplasma/parsec/parsec/utils/zone_malloc.c:41:33: error: expected expression
   41 |     gdata->lock               = PARSEC_ATOMIC_UNLOCKED;
      |                                 ^
/ccs/home/bouteilla/parsec/dplasma/parsec/parsec/include/parsec/sys/atomic-c11.h:219:32: note: expanded from macro 'PARSEC_ATOMIC_UNLOCKED'
  219 | #define PARSEC_ATOMIC_UNLOCKED ATOMIC_FLAG_INIT
      |                                ^
/opt/cray/pe/cce/17.0.0/cce-clang/x86_64/lib/clang/17/include/stdatomic.h:171:26: note: expanded from macro 'ATOMIC_FLAG_INIT'
  171 | #define ATOMIC_FLAG_INIT { 0 }
      |                          ^
1 error generated.
```

Suspected that we have a mish-mash of c11/stdatomic types, maybe a problem with include file order, need to dig.

devreal commented:
> Suspected that we have a mish-mash of c11/stdatomic types, maybe a problem with include file order, need to dig.

No, this is invalid C11. `ATOMIC_FLAG_INIT` can only be used for static initialization, not in a dynamic context.

bosilca committed:
Correctly initialize an unlock atomic lock.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

abouteiller committed:
gpu-datain: when we are executing a CPU body we need to retrieve the CPU
copy if we are passed-in a GPU copy, and we need to retain/release the
copies that we are swapping

abouteiller committed:
gpu-datain: pure distribution collections have no data_of/data_of_key

abouteiller commented:
Now passing 1-gpu/node, 8 ranks PTG POTRF
Sorry I had to force-push there were issues with rebasing on master

abouteiller commented:
` UCX_HANDLE_ERRORS=freeze  PMIX_MCA_psec='' srun -wleconte -N1 -n1  --cpu-bind=v,ldoms --gpus-per-task=2 --pty gdb --args tests/testing_dpotrf  -N $((10*1024)) -t $((8*256)) --nruns 1 -x -v=2  -g 2 -p 1  -c 4    -- --mca bind_threads 1`

```
#4  0x00007fff9e7c9386 in __assert_fail () from /lib64/libc.so.6
#5  0x00007ffff4211481 in parsec_device_kernel_pop (gpu_device=0xee94b0, gpu_task=0x7ff048002ea0, gpu_stream=0xee9968)
    at /home/bouteill/parsec/dplasma-master/parsec/parsec/mca/device/device_gpu.c:2293
#6  0x00007ffff4210259 in parsec_device_progress_stream (gpu_device=0xee94b0, stream=0xee9968, progress_fct=0x7ffff4210f28 <parsec_device_kernel_pop>,
    task=0x7ff048002ea0, out_task=0x7ffffffe86e0) at /home/bouteill/parsec/dplasma-master/parsec/parsec/mca/device/device_gpu.c:2009
#7  0x00007ffff4212b41 in parsec_device_kernel_scheduler (module=0xee94b0, es=0x95f250, _gpu_task=0x3a529a0)
    at /home/bouteill/parsec/dplasma-master/parsec/parsec/mca/device/device_gpu.c:2706
#8  0x00007ffff56c5cac in hook_of_dpotrf_U_potrf_dtrsm_CUDA (es=0x95f250, this_task=0x7ff044006f00)
    at /home/bouteill/parsec/dplasma-master/build.cuda/src/dpotrf_U.c:6018
#9  0x00007ffff41e54ea in __parsec_execute (es=0x95f250, task=0x7ff044006f00) at /home/bouteill/parsec/dplasma-master/parsec/parsec/scheduling.c:182
#10 0x00007ffff41e617d in __parsec_task_progress (es=0x95f250, task=0x7ff044006f00, distance=1) at /home/bouteill/parsec/dplasma-master/parsec/parsec/scheduling.c:503
#11 0x00007ffff41e6cf5 in __parsec_context_wait (es=0x95f250) at /home/bouteill/parsec/dplasma-master/parsec/parsec/scheduling.c:794
#12 0x00007ffff41e72d0 in parsec_context_wait (context=0xea6a30) at /home/bouteill/parsec/dplasma-master/parsec/parsec/scheduling.c:1000
#13 0x0000000000405d53 in main (argc=46, argv=0x7ffffffea1a8) at /home/bouteill/parsec/dplasma-master/build.cuda/tests/testing_dpotrf.c:74
(gdb) list
2288                                         gpu_device->super.device_index, gpu_device->super.name,
2289                                         parsec_task_snprintf(tmp, MAX_TASK_STRLEN, this_task),
2290                                         gpu_copy, gpu_copy->super.super.obj_reference_count,
2291                                         i, original, current_readers);
2292                }
2293                assert(current_readers >= 0);
2294                if( (0 == current_readers) && !(flow->flow_flags & PARSEC_FLOW_ACCESS_WRITE) ) {
2295                     PARSEC_DEBUG_VERBOSE(20, parsec_gpu_output_stream,
2296                                         "GPU[%d:%s]:\tMake read-only copy %p [ref_count %d] available on flow %s",
2297                                         gpu_device->super.device_index, gpu_device->super.name, gpu_copy, gpu_copy->super.super.obj_reference_count, flow->name);
```

abouteiller committed:
Use lock initializers instead of static temps

abouteiller committed:
gpu-datain: add a configure option and change some namings

therault committed:
When managing CPU-only tasks that received a GPU data copy, skip output-only flows, for which checking if they are control flows segfaults

G-Ragghianti commented:
I think we need to create a CI test that targets gpu_nvidia and issues the job to that runner, correct?

abouteiller commented:
Failure in `stress` (and similar in `nvlink`) due to the code generating a pushback event when transferring the last tile between the GEMM -> DISCARD_C flow (m >= mt+1). This tile has no original->device_copies[0] because it was created directly without a backing DC (from a NEW in MAKE_C). 

```
d@00000 GPU[1:cuda(0)]: Retrieve data (if any) for GEMM(79, 0, 0)[79, 0, 0]<0> keys = {4f, f000000000000001, 4f} {tp: 2} @parsec_device_kernel_scheduler:2719
d@00000 GPU[1:cuda(0)]: Try to Pop GEMM(79, 0, 0)[79, 0, 0]<0> keys = {4f, f000000000000001, 4f} {tp: 2} @parsec_device_kernel_pop:2264
d@00000 GPU[1:cuda(0)]: read copy 0x7ff06462f970 [ref_count 1] on flow A has readers (1) @parsec_device_kernel_pop:2323
d@00000 GPU[1:cuda(0)]: read copy 0x7ff064002c10 [ref_count 2] on flow C has readers (0) @parsec_device_kernel_pop:2323
d@00000 GPU[1:cuda(0)]: OUT Data copy 0x7ff064002c10 [ref_count 2] for flow C @parsec_device_kernel_pop:2330
Process 2891337 stopped
* thread #11, name = 'stress', stop reason = signal SIGSEGV: address not mapped to object (fault address: 0x70)
    frame #0: 0x00007ffff7eaa89b libparsec.so.4`parsec_device_kernel_pop(gpu_device=0x0000555555f7b7b0, gpu_task=0x00007ff06462a8c0, gpu_stream=0x0000555555f7bc68) at device_gpu.c:2341:17
   2338             if( gpu_task->pushout & (1 << i) ) {
   2339                 /* TODO: make sure no readers are working on the CPU version */
   2340                 original = gpu_copy->original;
-> 2341                 PARSEC_DEBUG_VERBOSE(10, parsec_gpu_output_stream,
   2342                                     "GPU[%d:%s]:\tMove D2H data <%s:%x> copy %p [ref_count %d] -- D:%p -> H:%p requested",
   2343                                     gpu_device->super.device_index, gpu_device->super.name, flow->name, original->key, gpu_copy, gpu_copy->super.super.obj_reference_count,
   2344                                      (void*)gpu_copy->device_private, original->device_copies[0]->device_private);
```

Potential fix is to allocate a dev0copy like is done for the network received tiles, not sure why it doesn't already.

abouteiller committed:
mpigpu: fix a number of crashes in release_self_contained_data

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -733,12 +732,22 @@ static void parsec_device_memory_release_list(parsec_device_gpu_module_t* gpu_de
+                           gpu_device->super.device_index, gpu_device->super.name, original->key, (void*)gpu_copy->device_private);
         }
-        assert(0 != (gpu_copy->flags & PARSEC_DATA_FLAG_PARSEC_OWNED) );
+        assert(0 != (gpu_copy->flags & PARSEC_DATA_FLAG_PARSEC_OWNED));
```
I know this has been here before, but: how should we handle copies that are not owned by parsec? They still seem to end up in the LRU even if they are only managed, not owned.

devreal commented:
I get this error with TTG when constraining the device memory to 10%. The copy is being pushed out when we try to change ownership to the device. 

```
testing_dpotrf_hip-parsec: /ccs/home/jschuchart/src/ttg/ttg/build_frontier/_deps/parsec-src/parsec/data.c:423: parsec_data_start_transfer_ownership_to_copy: Assertion `data->device_copies[i]->data_transfer_status != PARSEC_DATA_STATUS_UNDER_TRANSFER' failed.
```

devreal commented on parsec/data.c:
```diff
@@ -565,3 +590,42 @@ parsec_data_destroy( parsec_data_t *data )
+    for( uint32_t i = 0; i < parsec_nb_devices; i++) {
+        if (NULL == (copy = data->device_copies[i])) continue;
+        assert(1 == copy->super.super.obj_reference_count && 0 == copy->readers);
+        if( 0 == copy->device_index ) {
```
Is this check sufficient to handle the non-GPU devices (like recursive)?

abouteiller replied:
updated to use the device_is_gpu function

abouteiller commented on parsec/mca/device/device_gpu.c:
```diff
@@ -2350,45 +2478,60 @@ parsec_device_kernel_epilog( parsec_device_gpu_module_t *gpu_device,
+            gpu_copy->coherency_state = PARSEC_DATA_COHERENCY_SHARED;
+            assert(PARSEC_DATA_STATUS_UNDER_TRANSFER == cpu_copy->data_transfer_status);
+            cpu_copy->data_transfer_status = PARSEC_DATA_STATUS_COMPLETE_TRANSFER;
+            if( 0 == (parsec_mpi_allow_gpu_memory_communications & PARSEC_RUNTIME_SEND_GPU_MEMORY) ) {
```
Issue with stress and friends is stemming from here: when we pushout and we have a successor task, the successor task must receive the cpu copy as input (otherwise it will reference a gpu_copy that is now in the read LRU)

bosilca replied:
if a pushout is requested then we should always pass back the CPU copy (after it was updated). I need to understand the case yo are describing here. Is there a simple reproducer I can play with ?

abouteiller replied:
The reproducer is the stress CI tester

abouteiller committed:
testing for dev_index==0 is not always true, use the mca_device_is_gpu
accessor

abouteiller commented:
> I merged with master but apparently I missed a defect in erroneous cases printouts, that causes the CI failures.

ci failure due to using data_in uninitialized in new context nc during ontask, introduced by changes to task_snprintf in 108b7781b

abouteiller commented:
> I get this error with TTG when constraining the device memory to 10%. The copy is being pushed out when we try to change ownership to the device.
> 
> ```
> testing_dpotrf_hip-parsec: /ccs/home/jschuchart/src/ttg/ttg/build_frontier/_deps/parsec-src/parsec/data.c:423: parsec_data_start_transfer_ownership_to_copy: Assertion `data->device_copies[i]->data_transfer_status != PARSEC_DATA_STATUS_UNDER_TRANSFER' failed.
> ```

Is this resolved with #733 ?

devreal commented:
I see dangling copies on the device. This might just require a fix in the cleanup code (ignoring data that only has the device copy):

```
[****] TIME(s)      7.35572 : dpotrf	PxQxg=   2 2   2 NB= 2048 N=  131072 :  102044.227028 gflops - ENQ&PROG&DEST      7.75205 :   96827.061209 gflops - ENQ      0.39626 - DEST      0.00008
W@00002 GPU[2:hip(1)] still OWNS the master memory copy for data 0 (0x7fe11ca00000) and it is discarding it!
[****] TIME(s)      6.82278 : dpotrf	PxQxg=   2 2   2 NB= 2048 N=  131072 :  110015.002854 gflops - ENQ&PROG&DEST      6.82296 :  110012.229403 gflops - ENQ      0.00009 - DEST      0.00008
[****] TIME(s)      6.70474 : dpotrf	PxQxg=   2 2   2 NB= 2048 N=  131072 :  111951.924658 gflops - ENQ&PROG&DEST      6.70490 :  111949.236353 gflops - ENQ      0.00009 - DEST      0.00007
W@00002 GPU[2:hip(1)] still OWNS the master memory copy for data 0 (0x7fded0a00000) and it is discarding it!
[****] TIME(s)      6.64373 : dpotrf	PxQxg=   2 2   2 NB= 2048 N=  131072 :  112979.994900 gflops - ENQ&PROG&DEST      6.64388 :  112977.490061 gflops - ENQ      0.00009 - DEST      0.00006
W@00002 GPU[2:hip(1)] still OWNS the master memory copy for data 0 (0x7fe11aa00000) and it is discarding it!
W@00002 GPU[2:hip(1)] still OWNS the master memory copy for data 0 (0x7fe138a00000) and it is discarding it!
[****] TIME(s)      6.71053 : dpotrf	PxQxg=   2 2   2 NB= 2048 N=  131072 :  111855.299517 gflops - ENQ&PROG&DEST      6.71068 :  111852.904738 gflops - ENQ      0.00009 - DEST      0.00005
```

devreal committed:
Use PRIx64 when printing a parsec_data_key_t

Mismatch in printf type sizes might lead to segmentation faults.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

bosilca commented:
The original associated with these device owned copies should not have a valid dc ?

devreal commented:
Here is what I think happens in the `stress` benchmark:

- We allocate a new C tile, which has a host and device copy.
- At the end of the GEMM task, the device copy gets passed through the reshape code and a reference is added to the device copy because that is the copy that is captured in the `parsec_dep_data_description_t`: 
```C
    data.data   = this_task->data._f_A.data_out;
```
- We release the host copy at the end of `release_deps_of_stress_GEMM`. However, the host copy only has a single reference (the data) and so the host copy is detached from the `parsec_data_t`. Next time we use that `parsec_data_t` (in the next GEMM) we are missing the host copy. 

I don't understand the reshape code and I was hoping to never have to touch it. I suspect that the reshape code was not designed with GPUs in mind but I could be wrong. Will need some help digging through this and figuring out

1) Whether capturing the device copy in reshape is correct.
2) How we can make sure that the host copy is not accidentally released.

bosilca committed:
Allow command line arguments for the stress test.

Add missing dependency library.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Always propagate the CPU copy to successors.

This disable the communications from GPU memory, but it is necessary for
a proper tracking of reference counts on the data copies. At this point
I dont think we can fiddle with the code to allow for device copy
propagation, too many corner cases to be able to be workable.

Instead, we should rethink the entire data copy framework, and allow
tasks to fetch their inputs from devices as needed.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca commented:
For completeness let me try to describe what the issue was. In fact @abouteiller pointed to it earlier in the comments, but it went unnoticed.

To clarify: in order to allow sending to remote successors directly from the GPU, this PR made a drastic change by passing to successors the data copy the task has worked on, instead of the CPU version. However, it did not allowed the two versions to drift apart, it updated the CPU version to mirror the device copy (which is great as the two are now equivalent). But by passing the device data copy it alters the way we track the reference count for copies (propagate the output copies to successors to trigger their ref count update before releasing the refcounts on the current task input). In most cases, it works just fine because the CPU copies has other references to it, at least because it belongs to a data collection.

But in this test we are using a NEW data copy, one that is not associated with any data collection but instead specifically created for the chain of GEMM task. The first time we create this data, we also create the CPU mirror for it, but that mirrored copy has a refcount of 1 because it is only referenced once (ir does not belong to a data collection). So, when the first GEMM task in the chain passes the GPU copy to its successor and releases it's own input copies, it releases the only reference to the CPU copy allowing the runtime to dispose of the copy. This is bad because now we have a valid device copy passing around tasks that does not have a backend CPU copy.

bosilca committed:
Set the data_in to NULL before using the temporary task.

I'm not sure I understand how this worked so far, because the temporary
stack-based task we use to prepare the successors is barely initialized,
not enough to call parsec_task_snprintf on it. The least info we need to
set is to NULLify the data_in for all task classs flows.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Dont use the copy before it was allocated.

Fixes typos.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Fix typos

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

devreal commented:
I have run the new version on Frontier. I see a 4x slowdown and a Segfault at 64 nodes.
I don't have the backtraces for the Segfault but since the slowdown is significant I don't we think we want to run with this solution anyway.

`DPLASMA` here refers to this branch before the latest changes, while `new cudacomm` refers to the current HEAD of this branch.

<img width="992" height="602" alt="image" src="https://github.com/user-attachments/assets/f014642e-8dc5-4156-9f6b-e950542724a5" />

bosilca commented:
The performance drop is expected, one of my last commits disable the GPU-direct sending of the data. I need to revisit this, but in the current codebase sending or receiving data directly from devices might be very challenging to do.

bosilca committed:
arena: harden allocation failure paths

Handle arena allocation failures through shared cleanup paths.  When
parsec_arena_get_chunk() or parsec_arena_allocate_device_private() fails
after accounting memory in arena->used, jump to a common failure label that
logs the reason and atomically rolls back the used count before returning.

This avoids leaving arena accounting inflated when allocation is refused by
max_used or when the allocator returns NULL, and makes the failure visible
through PARSEC_DEBUG_VERBOSE.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
ptg: keep generated taskpool destructors from releasing borrowed arenas

Align the PTG-generated taskpool destructor with the arena datatype ownership
model from current master.  Generated taskpools can receive arena datatypes by
shallow copy from user code, so they must not release the embedded arena
pointers during taskpool destruction.

This avoids leaving the original arena datatype objects with stale arena
pointers when multiple taskpools reuse the same arena datatypes, as exposed by
the reshape tests.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
gpu: protect CPU mirrors for GPU-aware sends

Allow GPU-aware remote dependencies to keep propagating the GPU copy when the
communication layer can send directly from GPU memory, instead of always
forcing the task output back to the CPU copy.

For self-contained temporaries, mark the CPU copy as protected while GPU copies
are being propagated.  The data copy release path now preserves that final CPU
mirror until the complete self-contained data object can be released, and the
GPU LRU detach paths trigger that cleanup once protected GPU copies are evicted
or reused.

Also include the protected flag in data copy debug dumps and avoid releasing an
arena chunk from a copy whose arena allocation did not complete.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
gpu: keep non-owned device copies out of PaRSEC LRUs

Non-owned GPU data copies can be used directly as task inputs, but their
lifetime is not controlled by PaRSEC. Keep those copies out of the GPU memory
LRU when their reader count drops to zero.

When a non-owned copy is reused as a read input, still account for the read in
kernel_push and release it in kernel_pop. This keeps reader accounting balanced
without making PaRSEC responsible for availability, write-back, or reclamation
of externally owned memory.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
remote_dep: default receive placement before datatype early exits

Initialize the preferred device for incoming remote dependencies before any
early return in the datatype retrieval callback. In particular, the packed
fallback for mixed receive datatypes should keep the default CPU placement
instead of inheriting a stale GPU hint from a recycled remote_deps item.

Also clear the preferred device when recycling remote dependency outputs so
future activations start from a deterministic CPU default.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
gpu: honor per-tile allocation for self-contained copies

Allocate GPU-backed arena temporaries with the device allocator when
PARSEC_GPU_ALLOC_PER_TILE is enabled, and release self-contained GPU copies
through the matching device free path.

Keep the existing zone allocator path for pooled GPU-memory builds.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
device/gpu: preserve input-stream ordering for pending transfers

Restore event skipping for truly empty GPU stream stages while still recording
an input-stream event when a task depends on an input copy already marked
UNDER_TRANSFER.  The stage-in path now reports such already-queued transfers as
input stream work, so stream ordering prevents the task from executing before
the earlier transfer completes.

Also keep successful kernel submissions on the event-record path by returning a
positive value from the execution-stage wrapper, and clean up the debug message
for inputs that are already resident on the target GPU.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
bug hunting

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
tests: share RTT PTG between pingpong and CUDA

Unify the pingpong and CUDA RTT tests around a single shared
tests/apps/pingpong/rtt.jdf. Add fragment support to the shared RTT PTG,
provide both CPU and CUDA bodies, and define rtt_verbose in the generated
JDF source with wrappers importing it.

Keep the tests' separate behavior in their drivers: the pingpong app runs
one CPU-only fragment chain, while cuda_rtt passes the requested fragment
count and restricts execution to CUDA devices. Update the pingpong custom
descriptor to accept A(f, rank), and point the CUDA test CMake target at the
shared JDF while removing the duplicate runtime/cuda/rtt.jdf.

Document the RTT and bandwidth test intent in the JDF preambles and update
NVIDIA copyright notices in the modified test files.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
device/gpu: force pushout for remote GPU outputs without send support

When direct GPU sends are disabled, a GPU task output that feeds a remote
successor must be staged back to host memory before dependency release. Otherwise
the remote dependency path can receive a GPU datacopy and attempt to send its
device pointer even though the communication engine is running with
mpi_gpu_aware=0 or 1.

Add a late GPU-task pushout pass, run just before output retrieval, that preserves
pushout bits already set by the upper layer and only adds bits for write flows
with remote task successors. The extra successor walk is only enabled in
distributed builds and only when direct GPU sends are disabled, avoiding added
task submission latency and avoiding overhead for normal GPU-aware send modes.

This keeps final data-collection writebacks under the existing PTG/DTD pushout
logic while making remote task successors safe in degraded GPU communication
modes.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Fix the ptr_pingpong test.

Multiple tasks were writing into the same data copy. The compiler is
unable to flag this type of error, but somehow the CUDA driver did. The
logic was convoluted:
- the first task acquired the data copy and upon completion started to
  move out the content, marking the data UNDER_TRANSFER along the way
- a second task tried to acquire the data, but as it was under transfer
  it set its data_out to NULL.
- Later on the second task acquired the data copy again this time
  without the UNDER_TRANSFER flag but did not set the data_out anymore
  leading to a segfault in the pushout

Anyway, the test is incorrect as is, the flow should be marked a
readonly.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
parsec: normalize parsec_init option vectors internally

Treat parsec_init() arguments as a PaRSEC option vector instead of
requiring callers to provide a traditional process argv with argv[0] set
to the application name. The command-line parser still expects argv[0] to
be skipped, so parsec_init() now builds that internal shape itself before
parsing MCA options.

Keep compatibility with existing callers that pass the full main(argc,
argv), or a slice starting with "--", by ignoring a leading non-option
token or leading "--" separator.

This fixes tests that forward only the PaRSEC option tail, where an
initial "--mca" argument could previously be swallowed as argv[0]. In
particular, CUDA redistribute tests can now override the default
PARSEC_MCA_device_cuda_enabled=0 environment set by CTest.

Update the parsec_init() API documentation to clarify that pargv contains
PaRSEC options and does not reserve an application-name slot.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
device/gpu: preserve GPU write dependencies under transfer

Handle GPU task readiness when a writable dependency is already resident on
the target GPU but still marked UNDER_TRANSFER. Keep the selected data_out
binding in place, return AGAIN, and retry after transfer completion bumps the
device data availability epoch instead of clearing the binding and losing the
selected copy.

Record an input-stream event for already-resident inputs that are still under
transfer so stream ordering prevents execution before the queued transfer is
complete. Bump data_avail_epoch when input transfers complete to avoid
livelock in the reserve-space fast path.

When a GPU task must push data back to host memory, lazily recreate a missing
CPU mirror for self-contained GPU data by querying the successor input
datatype and allocating the host copy from the returned arena. Background D2H
cleanup tasks cannot recover this metadata, so leave GPU-only copies for the
real successor pushout path.

Keep speculative GPU reservations in temp_loc[] while reserve_space scans all
flows, and use that list directly to return only newly reserved copies to the
LRU if the task must retry because a writable copy is still under transfer.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca commented:
This PR grew too much. While there is a Major Theme in it, aka. making the GPU support great again, a lot of side themes crept in. 

1. GPU-aware communication support
  The branch adds infrastructure for remote dependencies to carry GPU data copies instead of always collapsing back to CPU/device 0. It introduces GPU-aware communication configuration and adjusts remote dependency paths so the correct device copy can be propagated.

2. Safer GPU/CPU copy lifetime management
A lot of the branch is about preventing stale, leaked, or prematurely released copies:
 - protect CPU mirrors when GPU-aware sends use GPU copies;
 - keep non-owned GPU copies out of PaRSEC LRUs;
 - honor per-tile GPU allocation/freeing for self-contained data;
 - avoid releasing borrowed or incomplete arena-backed copies incorrectly.

3. Correct behavior when direct GPU sends are unavailable
The branch adds a late pushout pass for remote GPU outputs when mpi_gpu_aware=0 or 1, so remote successors get a valid host copy instead of accidentally sending a GPU pointer through a non-GPU-capable path.

4. Transfer ordering and UNDER_TRANSFER correctness
The newer GPU commits harden the state machine around in-flight transfers:
 - tasks retry instead of losing data_out when a writable GPU dependency is still UNDER_TRANSFER;
 - input stream events are recorded when a dependency is already under transfer;
 - transfer completion bumps the GPU data availability epoch;
 - missing CPU mirrors for self-contained GPU data are lazily reconstructed from successor datatype/arena metadata.
 
5. Arena and PTG robustness fixes
The branch hardens arena allocation failure cleanup and fixes generated PTG destructor behavior so borrowed arenas are not released by generated taskpools. It also allows JDFs with no dependencies/datatypes/arenas.

6. Test improvements
The branch adds and refactors CUDA communication tests:
 - adds a CUDA RTT test;
 - unifies pingpong and CUDA RTT around a shared tests/apps/pingpong/rtt.jdf;
 - adds fragments/verbosity support;
 - fixes ptg_pingpong;
 - improves CUDA stress-test argument handling.
 - parsec_init() option handling

7. parsec_init() now treats its argument vector as PaRSEC options, without requiring callers to fake an application argv[0]. This fixes tests that pass only MCA option tails, especially CUDA tests relying on device MCA overrides.

However, if there is one thing to take away from it is that PaRSEC now has full GPU support for the communication engine (driven by MCA params), and full support for NEW data on the GPU and outside. As a bonus, we can now create a new data on the GPU and then transfer it to the CPU, and the backend CPU copy will be created lazily. A new era for PaRSEC's GPU support.
